File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
from collections import defaultdict, deque

# 그래프를 인접 리스트로 표현
graph = defaultdict(list)

# 입력 받기
N, M, V = map(int, input().split())
for _ in range(M):
    u, v = map(int, input().split())
    graph[u].append(v)
    graph[v].append(u)

# 각 정점의 인접 리스트를 정렬 (정점 번호가 작은 것을 먼저 방문하기 위해)
for key in graph:
    graph[key].sort()

def dfs(graph, start):
    visited = []
    stack = [start]
    
    while stack:
        node = stack.pop()
        if node not in visited:
            visited.append(node)
            # 방문할 수 있는 정점이 여러 개인 경우, 정점 번호가 작은 것을 먼저 방문하기 위해 정렬
            stack.extend(sorted(graph[node], reverse=True))
    return visited

def bfs(graph, start):
    visited = []
    queue = deque([start])
    
    while queue:
        node = queue.popleft()
        if node not in visited:
            visited.append(node)
            # 방문할 수 있는 정점이 여러 개인 경우, 정점 번호가 작은 것을 먼저 방문하기 위해 정렬
            queue.extend(sorted(graph[node]))
    return visited

# DFS와 BFS 수행
dfs_result = dfs(graph, V)
bfs_result = bfs(graph, V)

# 결과 출력
print(' '.join(map(str, dfs_result)))
print(' '.join(map(str, bfs_result)))

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.js
let input = require('fs').readFileSync('/dev/stdin').toString().trim().split('\n');
let [N, M, V] = input[0].split(' ').map(Number);
let graph = Array.from({ length: N + 1 }, () => []);

for(let i=1; i<=M; i++){
    let [a, b] = input[i].split(' ').map(Number);
    graph[a].push(b);
    graph[b].push(a);
}
graph.forEach(adj => adj.sort((a,b) => a-b));

//DFS
function DFS(start){
    let visited = [];
    let stack = [start];
    let visitedSet = new Set(); 

    while(stack.length){
        let node = stack.pop();
        if(!visitedSet.has(node)){
            visited.push(node);
            visitedSet.add(node);
            stack.push(...graph[node].slice().reverse());
        }
    }
    return visited;
}

function BFS(start){
    let visited = [];
    let queue = [start];
    let visitedSet = new Set();
    
    while(queue.length){
        let node = queue.shift();
        if(!visitedSet.has(node)){
            visited.push(node);
            visitedSet.add(node);
            queue.push(...graph[node]);
        }
    }
    return visited;
}
console.log(DFS(V).join(' '));
console.log(BFS(V).join(' '));
    
  